File: inference/models/mpt.cc
/* Copyright 2023 CMU, Facebook, LANL, MIT, NVIDIA, and Stanford (alphabetical)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "mpt.h"

namespace FlexFlow {

using namespace Legion;
using json = nlohmann::json;

void MPT::create_mpt_model(FFModel &ff,
                           std::string const &model_config_file_path,
                           std::string const &weight_file_path,
                           InferenceMode mode,
                           GenerationConfig generationConfig,
                           bool use_full_precision) {
  MPTConfig mpt_config(model_config_file_path);
  mpt_config.print();

  if (ff.config.tensor_parallelism_degree > mpt_config.n_heads ||
      mpt_config.n_heads % ff.config.tensor_parallelism_degree != 0) {
    assert(false && "The number of attention heads is smaller, or it is not "
                    "divisible by the tensor parallelism degree");
  }

  std::unordered_map<std::string, Layer *> weights_layers;

  //------------------------------ build the model --------------------------
  Tensor input;
  {
    int const token_dims[] = {
        (mode == TREE_VERIFY_MODE || mode == BEAM_SEARCH_MODE)
            ? BatchConfig::max_verify_tokens_per_batch()
            : BatchConfig::max_tokens_per_batch(),
        1};
    input = ff.create_tensor<2>(token_dims, DT_INT32);
  }

  Initializer *embed_init = new UniformInitializer(std::rand(), 0, 0);
  std::vector<int> axes = {0};

  Tensor hidden_states = ff.embedding(input,
                                      mpt_config.vocab_size,
                                      mpt_config.hidden_size,
                                      AGGR_MODE_NONE,
                                      use_full_precision ? DT_FLOAT : DT_HALF,
                                      NULL,
                                      embed_init,
                                      "wte");

  Tensor intermediate_output = nullptr, layernorm_output = nullptr;
  Tensor res_ln_outputs[2] = {nullptr, nullptr};

  for (int i = 0; i < mpt_config.n_layers; i++) {
    ff.set_transformer_layer_id(i);

    if (i == 0) {
      layernorm_output = ff.layer_norm(
          hidden_states,
          axes,
          true,
          1e-05,
          false,
          DT_NONE,
          std::string("layers." + std::to_string(i) + ".norm_1").c_str());
    } else {
      ff.residual_layer_norm(
          intermediate_output,
          hidden_states,
          nullptr,
          res_ln_outputs,
          false,
          axes,
          true,
          1e-05,
          false,
          false,
          DT_NONE,
          std::string("layers." + std::to_string(i) + ".norm_1").c_str());
      hidden_states = res_ln_outputs[0];
      layernorm_output = res_ln_outputs[1];
    }

    Tensor qkv_proj = ff.dense(
        layernorm_output,
        mpt_config.hidden_size *
            3, // q, k, v. need to change if want to remove replication.
               // (q_heads + 2 * kv_heads) * proj_size
        AC_MODE_NONE,
        false,         // seems like it does not use bias
        DT_NONE,       // what is this
        nullptr,       // ?
        nullptr,       // ?
        nullptr,       // ?
        REG_MODE_NONE, // no regularization
        0.0f,          // no dropout
        std::string("layers." + std::to_string(i) + ".attn.qkv_proj").c_str());

    Tensor o_proj;
    switch (mode) {
      case BEAM_SEARCH_MODE: {
        o_proj = ff.spec_inc_multihead_self_attention(
            qkv_proj,
            mpt_config.hidden_size,
            mpt_config.n_heads,
            mpt_config.hidden_size / mpt_config.n_heads,
            mpt_config.hidden_size / mpt_config.n_heads,
            0.0f,
            false,
            false,
            false,
            DT_NONE, /*data_type*/
            NULL,
            mpt_config.rotary_embedding_meta,
            /*scaling query*/ true,
            /*scaling factor*/
            pow((mpt_config.hidden_size / mpt_config.n_heads), -0.5),
            /*qk_prod_scaling*/ false,
            /*position_bias*/ true,
            std::string("layers." + std::to_string(i) + ".attn")
                .c_str() /*name*/
        );
        break;
      }
      case TREE_VERIFY_MODE: {
        o_proj = ff.inc_multihead_self_attention_verify(
            qkv_proj,
            mpt_config.hidden_size,
            mpt_config.n_heads,
            mpt_config.hidden_size / mpt_config.n_heads,
            mpt_config.hidden_size / mpt_config.n_heads,
            0.0f,
            false,
            false,
            false,
            DT_NONE, /*data_type*/
            NULL,
            mpt_config.rotary_embedding_meta,
            /*scaling query*/ true,
            /*scaling factor*/
            pow((mpt_config.hidden_size / mpt_config.n_heads), -0.5),
            /*qk_prod_scaling*/ false,
            /*position_bias*/ true,
            std::string("layers." + std::to_string(i) + ".attn")
                .c_str() /*name*/
        );
        break;
      }
      case INC_DECODING_MODE: {
        o_proj = ff.inc_multihead_self_attention(
            qkv_proj,
            mpt_config.hidden_size,
            mpt_config.n_heads,
            mpt_config.hidden_size / mpt_config.n_heads,
            mpt_config.hidden_size / mpt_config.n_heads,
            0.0f,
            false,
            false,
            false,
            DT_NONE, /*data_type*/
            NULL,
            mpt_config.rotary_embedding_meta,
            /*scaling query*/ true,
            /*scaling factor*/
            pow((mpt_config.hidden_size / mpt_config.n_heads), -0.5),
            /*qk_prod_scaling*/ false,
            /*position_bias*/ true,
            std::string("layers." + std::to_string(i) + ".attn")
                .c_str() /*name*/
        );
        break;
      }
      default: {
        assert(false);
      }
    }

    Tensor attn_outputs = ff.dense(
        o_proj,
        mpt_config.hidden_size,
        AC_MODE_NONE,
        false,
        DT_NONE,
        nullptr,
        nullptr,
        nullptr,
        REG_MODE_NONE,
        0.0f,
        std::string("layers." + std::to_string(i) + ".attn.o_proj").c_str());

    ff.residual_layer_norm(
        attn_outputs,
        hidden_states,
        nullptr,
        res_ln_outputs,
        false,
        axes,
        true,
        1e-05,
        false,
        false,
        DT_NONE,
        std::string("layers." + std::to_string(i) + ".norm_2").c_str());
    hidden_states = res_ln_outputs[0];
    layernorm_output = res_ln_outputs[1];

    // MLP
    layernorm_output = ff.dense(
        layernorm_output,
        4 * mpt_config.hidden_size,
        AC_MODE_NONE,
        false,
        DT_NONE,
        nullptr,
        nullptr,
        nullptr,
        REG_MODE_NONE,
        0.0f,
        std::string("layers." + std::to_string(i) + ".ffn.up_proj").c_str());
    layernorm_output = ff.gelu(layernorm_output);
    intermediate_output = ff.dense(
        layernorm_output,
        mpt_config.hidden_size,
        AC_MODE_NONE,
        false,
        DT_NONE,
        nullptr,
        nullptr,
        nullptr,
        REG_MODE_NONE,
        0.0f,
        std::string("layers." + std::to_string(i) + ".ffn.down_proj").c_str());
  }

  // final
  ff.residual_layer_norm(intermediate_output,
                         hidden_states,
                         nullptr,
                         res_ln_outputs,
                         false,
                         axes,
                         true,
                         1e-05,
                         false,
                         false,
                         DT_NONE,
                         "norm_f");
  Tensor all_final_norm = res_ln_outputs[1];

  Tensor lm_head = ff.dense(all_final_norm,
                            mpt_config.vocab_size,
                            AC_MODE_NONE,
                            false,
                            DT_NONE,
                            nullptr,
                            nullptr,
                            nullptr,
                            REG_MODE_NONE,
                            0.0f,
                            "lm_head");

  Tensor output;
  if (mode == BEAM_SEARCH_MODE) {
    Tensor softmax = ff.softmax(lm_head, -1);
    output = ff.argmax(softmax, /*beam_Search*/ true);
  } else {
    output = ff.argmax(lm_head, /*beam_Search*/ false);
  }
  FileDataLoader *fileloader =
      new FileDataLoader("",
                         weight_file_path,
                         mpt_config.n_heads,
                         mpt_config.n_heads,
                         mpt_config.hidden_size,
                         mpt_config.hidden_size / mpt_config.n_heads,
                         ff.config.tensor_parallelism_degree,
                         use_full_precision);

  InferenceManager *im = InferenceManager::get_inference_manager();
  im->register_model_weights_loader(&ff, fileloader);
}

}; // namespace FlexFlow

File: python/flexflow/serve/models/mpt.py
# Copyright 2023 CMU, Facebook, LANL, MIT, NVIDIA, and Stanford (alphabetical)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from flexflow.core import *
from .base import FlexFlowModel
import random, torch, shutil


class MPTConfig:
    def __init__(self, hf_config):
        self.max_beam_width = 1
        self.max_beam_depth = 8
        self.max_spec_tree_token_num = 20
        self.hidden_size = hf_config.d_model
        self.n_heads = hf_config.n_heads
        self.n_layers = hf_config.n_layers
        self.vocab_size = hf_config.vocab_size
        self.rotary_embedding_meta = RotaryEmbeddingMeta(apply_rotary_embedding=False)
        # Standardized FlexFlow num heads fields below
        self.num_attention_heads = hf_config.n_heads
        self.num_key_value_heads = hf_config.n_heads


class FlexFlowMPT(FlexFlowModel):
    def __init__(
        self,
        mode,
        generation_config,
        ffconfig,
        hf_config,
        data_type,
        # max_batch_size=1,
        # max_seq_length=256,
        max_tokens_per_batch,
        weights_filepath="",
        tokenizer_filepath="",
    ):
        self.mode = mode
        self.generation_config = generation_config
        self.ffconfig = ffconfig
        self.data_type = data_type
        self.mpt_config = MPTConfig(hf_config)
        self.weights_filepath = weights_filepath
        self.tokenizer_filepath = tokenizer_filepath
        self.maxint = 2**31 - 1
        max_verify_tokens_per_batch = (
            max_tokens_per_batch + self.mpt_config.max_spec_tree_token_num
        )

        # Sanity checks
        if self.mpt_config.hidden_size % self.mpt_config.n_heads != 0:
            raise ValueError(
                f"Hidden size ({self.mpt_config.hidden_size}) is not divisible by n_head ({self.mpt_config.n_heads})"
            )

        # Sanity checks
        if (
            self.mpt_config.n_heads < self.ffconfig.tensor_parallelism_degree
            or self.mpt_config.n_heads % self.ffconfig.tensor_parallelism_degree != 0
        ):
            raise ValueError(
                f"Number of attention heads ({self.mpt_config.n_heads}) is smaller, or not divisible by tensor parallelism degree ({self.ffconfig.tensor_parallelism_degree})"
            )
        self.build_model(
            max_tokens_per_batch
            if self.mode == InferenceMode.INC_DECODING_MODE
            else max_verify_tokens_per_batch
        )

    def build_model(self, max_tokens_per_batch):
        ffmodel = FFModel(self.ffconfig)

        tokens_dims = [max_tokens_per_batch, 1]
        input = ffmodel.create_tensor(tokens_dims, DataType.DT_INT32)

        embed_init = UniformInitializer(random.randint(0, self.maxint), 0, 0)
        hidden_states = ffmodel.embedding(
            input,
            self.mpt_config.vocab_size,
            self.mpt_config.hidden_size,
            AggrMode.AGGR_MODE_NONE,
            self.data_type,
            None,
            embed_init,
            name="wte",
        )

        axes = [
            0,
        ]

        for i in range(self.mpt_config.n_layers):
            ffmodel.set_transformer_layer_id(i)

            if i == 0:
                layernorm_output = ffmodel.layer_norm(
                    hidden_states,
                    axes,
                    True,
                    1e-05,
                    False,
                    name=f"layers.{i}.norm_1",
                )
            else:
                hidden_states, layernorm_output = ffmodel.residual_layer_norm(
                    intermediate_output,
                    hidden_states,
                    None,
                    False,
                    axes,
                    True,
                    1e-05,
                    False,
                    name=f"layers.{i}.norm_1",
                )

            qkv_proj = ffmodel.dense(
                layernorm_output,
                3 * self.mpt_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers.{i}.attn.qkv_proj",
            )

            if self.mode == InferenceMode.BEAM_SEARCH_MODE:
                o_proj = ffmodel.spec_inc_multihead_self_attention(
                    qkv_proj,
                    self.mpt_config.hidden_size,
                    self.mpt_config.n_heads,
                    self.mpt_config.hidden_size // self.mpt_config.n_heads,
                    self.mpt_config.hidden_size // self.mpt_config.n_heads,
                    0.0,  # dropout
                    False,  # qkv_bias
                    False,  # final_bias
                    False,  # add_zero_attn
                    DataType.DT_NONE,  # data_type
                    None,  # kernel initializer
                    self.mpt_config.rotary_embedding_meta,
                    True,  # scaling_query
                    (self.mpt_config.hidden_size / self.mpt_config.n_heads)
                    ** (-0.5),  # scaling_factor
                    False,  # qk_prod_scaling
                    True,  # qk_prod_scaling
                    name=f"layers.{i}.attn",
                )
            elif self.mode == InferenceMode.TREE_VERIFY_MODE:
                o_proj = ffmodel.inc_multihead_self_attention_verify(
                    qkv_proj,
                    self.mpt_config.hidden_size,
                    self.mpt_config.n_heads,
                    self.mpt_config.hidden_size // self.mpt_config.n_heads,
                    self.mpt_config.hidden_size // self.mpt_config.n_heads,
                    0.0,  # dropout
                    False,  # qkv_bias
                    False,  # final_bias
                    False,  # add_zero_attn
                    DataType.DT_NONE,  # data_type
                    None,  # kernel initializer
                    self.mpt_config.rotary_embedding_meta,
                    True,  # scaling_query
                    (self.mpt_config.hidden_size / self.mpt_config.n_heads)
                    ** (-0.5),  # scaling_factor
                    False,  # qk_prod_scaling
                    True,  # qk_prod_scaling
                    name=f"layers.{i}.attn",
                )
            elif self.mode == InferenceMode.INC_DECODING_MODE:
                o_proj = ffmodel.inc_multihead_self_attention(
                    qkv_proj,
                    self.mpt_config.hidden_size,
                    self.mpt_config.n_heads,
                    self.mpt_config.hidden_size // self.mpt_config.n_heads,
                    self.mpt_config.hidden_size // self.mpt_config.n_heads,
                    0.0,  # dropout
                    False,  # qkv_bias
                    False,  # final_bias
                    False,  # add_zero_attn
                    DataType.DT_NONE,  # data_type
                    None,  # kernel initializer
                    self.mpt_config.rotary_embedding_meta,
                    True,  # scaling_query
                    (self.mpt_config.hidden_size / self.mpt_config.n_heads)
                    ** (-0.5),  # scaling_factor
                    False,  # qk_prod_scaling
                    True,  # qk_prod_scaling
                    name=f"layers.{i}.attn",
                )
            else:
                assert False

            attn_outputs = ffmodel.dense(
                o_proj,
                self.mpt_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers.{i}.attn.o_proj"
            )

            hidden_states, layernorm_output = ffmodel.residual_layer_norm(
                attn_outputs,
                hidden_states,
                None,
                False,
                axes,
                True,
                1e-05,
                False,
                name=f"layers.{i}.norm_2",
            )
            # mlp
            layernorm_output = ffmodel.dense(
                layernorm_output,
                4 * self.mpt_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers.{i}.ffn.up_proj",
            )
            layernorm_output = ffmodel.gelu(layernorm_output)
            intermediate_output = ffmodel.dense(
                layernorm_output,
                self.mpt_config.hidden_size,
                ActiMode.AC_MODE_NONE,
                False,
                name=f"layers.{i}.ffn.down_proj",
            )

        _, all_final_norm = ffmodel.residual_layer_norm(
            intermediate_output,
            hidden_states,
            None,
            False,
            axes,
            True,
            1e-05,
            False,
            name=f"norm_f",
        )
        lm_head = ffmodel.dense(
            all_final_norm,
            self.mpt_config.vocab_size,
            ActiMode.AC_MODE_NONE,
            False,
            name="lm_head",
        )

        if self.generation_config.do_sample:
            dense = ffmodel.scalar_true_divide(
                lm_head, self.generation_config.temperature, False
            )
            softmax = ffmodel.softmax(dense, -1)
            output = ffmodel.sampling(softmax, self.generation_config.topp)
        else:
            softmax = ffmodel.softmax(lm_head, -1)
            output = ffmodel.argmax(softmax, False)

        self.ffmodel = ffmodel

    # TODO: finish this
    def convert_hf_weight_name(name):
        return (
            name.replace("transformer.blocks.", "layers.")
            .replace("transformer.", "")
            .replace("attn.out_proj", "attn.o_proj")
        )

    def convert_hf_model(model, dst_folder):
        os.makedirs(dst_folder, exist_ok=True)
        for name, params in model.named_parameters():
            name = FlexFlowMPT.convert_hf_weight_name(name)
            if "Wqkv" in name:
                name_q = name.replace("attn.Wqkv", "attn.q_proj")
                name_k = name.replace("attn.Wqkv", "attn.k_proj")
                name_v = name.replace("attn.Wqkv", "attn.v_proj")
                q, k, v = torch.split(
                    params,
                    [
                        model.config.d_model,
                        model.config.d_model,
                        model.config.d_model,
                    ],
                    0,
                )
                q.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_q))
                k.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_k))
                v.detach().cpu().numpy().tofile(os.path.join(dst_folder, name_v))
            else:
                params.detach().cpu().numpy().tofile(os.path.join(dst_folder, name))

        shutil.copy(
            os.path.join(dst_folder, "wte.weight"),
            os.path.join(dst_folder, "lm_head.weight"),
        )
